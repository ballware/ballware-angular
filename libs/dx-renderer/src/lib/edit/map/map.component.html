<div [class]="!((visible.visible$ | async) ?? false) ? 'hidden' : 'mt-2 dx-texteditor-label'" [style.position]="'relative'">{{livecycle.layoutItem?.options?.caption ?? ''}}</div>
<dx-map #map
  [visible]="(visible.visible$ | async) ?? false"
  [height]="livecycle.layoutItem?.options?.height ?? '100%'"
  [width]="'100%'"
  [apiKey]="{ google: (googlekey$ | async) ?? 'unlicensed' }"
  [provider]="'google'"
  [type]="'hybrid'"
  [autoAdjust]="true"
  [zoom]="30"
  [controls]="true">
</dx-map>