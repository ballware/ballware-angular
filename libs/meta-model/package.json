{
  "name": "@ballware/meta-model",
  "version": "8.0.1",
  "type": "commonjs"
}
