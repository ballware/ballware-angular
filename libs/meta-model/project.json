{
  "name": "meta-model",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/meta-model/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/libs/meta-model",
        "main": "libs/meta-model/src/index.ts",
        "tsConfig": "libs/meta-model/tsconfig.lib.json",
        "assets": [
          "libs/meta-model/*.md"
        ]
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs meta-model {args.ver} {args.tag}"
      },
      "dependsOn": [
        "build"
      ]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "libs/meta-model/**/*.ts"
        ]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}"
      ],
      "options": {
        "jestConfig": "libs/meta-model/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
