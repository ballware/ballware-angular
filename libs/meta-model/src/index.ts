export * from './lib/cruditem';
export * from './lib/document';
export * from './lib/entity';
export * from './lib/page';
export * from './lib/queryparams';
export * from './lib/rights';
export * from './lib/scriptutil';
export * from './lib/statistic';
export * from './lib/template';
export * from './lib/tenant';

