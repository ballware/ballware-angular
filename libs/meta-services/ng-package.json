{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/meta-services",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
