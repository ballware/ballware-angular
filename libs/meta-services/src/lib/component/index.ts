import { StoreModule } from '@ngrx/store';
import { componentReducer } from './component.reducer';
import { componentFeatureKey } from './component.state';

export * from './component.actions';
export * from './component.state';

export const ComponentFeatureModule = StoreModule.forFeature(componentFeatureKey, componentReducer);