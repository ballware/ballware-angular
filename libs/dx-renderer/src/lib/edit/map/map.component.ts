import { AfterViewInit, Component, Inject, ViewChild } from '@angular/core';
import { SETTINGS_SERVICE, SettingsService } from '@ballware/meta-services';
import { DxMapComponent, DxMapModule } from 'devextreme-angular';
import { Observable, combineLatest, takeUntil } from 'rxjs';
import { CommonModule } from '@angular/common';
import { Destroy, EditItemLivecycle, NullableLatLngValue, Readonly, Visible } from '@ballware/renderer-commons';

declare let google: any;

@Component({
  selector: 'ballware-edit-map',
  templateUrl: './map.component.html',
  styleUrls: [],
  imports: [CommonModule, DxMapModule],
  hostDirectives: [Destroy, { directive: EditItemLivecycle, inputs: ['initialLayoutItem'] }, NullableLatLngValue, Readonly, Visible],
  standalone: true
})
export class EditLayoutMapComponent implements AfterViewInit {

  @ViewChild('map', { static: false }) map?: DxMapComponent;

  public googlekey$: Observable<string|undefined>;

  constructor(
    @Inject(SETTINGS_SERVICE) private settingsService: SettingsService,
    public destroy: Destroy,
    public livecycle: EditItemLivecycle,
    public visible: Visible,
    public readonly: Readonly,
    public value: NullableLatLngValue
  ) {    
    this.googlekey$ = this.settingsService.googlekey$;
  }

  ngAfterViewInit(): void {
    combineLatest([this.readonly.readonly$, this.value.currentValue$])
      .pipe(takeUntil(this.destroy.destroy$))
      .subscribe(([readonly, value]) => {
        const existingMarkers = this.map?.instance.option(
          'markers'
        ) as Array<object>;

        for (const marker of existingMarkers) {
          this.map?.instance.removeMarker(marker);
        }

        if (value) {
          value.lat = value.lat ?? 0.0;
          value.lng = value.lng ?? 0.0;  
          
          this.map?.instance.addMarker({
            location: value
          }).then((marker) => {
            if (!readonly) {
              marker.setDraggable(true);

              google.maps.event.addListener(
                marker,
                'dragend',
                (e: {
                  latLng: { lng: () => number; lat: () => number };
                }) => {
                  this.value.value = {
                    lat: e.latLng.lat(),
                    lng: e.latLng.lng()
                  };
                }
              );
            }
          });
        }
      });
  }
}
