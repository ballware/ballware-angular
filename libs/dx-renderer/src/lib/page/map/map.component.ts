import { AfterViewInit, Component, Input, ViewChild } from '@angular/core';
import { CrudItem, EntityMapOptions, PageLayoutItem } from '@ballware/meta-model';
import { CrudService, SettingsService } from '@ballware/meta-services';
import { DxMapComponent } from 'devextreme-angular';
import { BehaviorSubject, Observable, combineLatest, takeUntil } from 'rxjs';
import { getByPath } from '../../utils/databinding';
import { DataSourceService } from '../../utils/datasource.service';
import { WithDestroy } from '../../utils/withdestroy';

@Component({
  selector: 'ballware-page-map',
  templateUrl: './map.component.html',
  styleUrls: ['./map.component.scss']
})
export class PageLayoutMapComponent extends WithDestroy() implements AfterViewInit {

  @Input() layoutItem?: PageLayoutItem;

  @ViewChild('map', { static: false }) map?: DxMapComponent;

  public googlekey$: Observable<string|undefined>;

  public markers$ = new BehaviorSubject<any[]>([]);

  private mouseTarget: Element|undefined|null;

  constructor(private settingsService: SettingsService, private crudService: CrudService, private dataSourceService: DataSourceService) {
    super();

    this.onMapMouseMove = this.onMapMouseMove.bind(this);
    this.onMarkerClicked = this.onMarkerClicked.bind(this);

    this.googlekey$ = this.settingsService.googlekey$;
  }

  ngAfterViewInit(): void {

    this.map?.instance.element().addEventListener('mousemove', this.onMapMouseMove);

    combineLatest([this.dataSourceService.dataSource$])
      .pipe(takeUntil(this.destroy$))
      .subscribe(([dataSource]) => {
        const locationMember = (this.layoutItem?.options?.itemoptions as EntityMapOptions)?.locationMember;

        if (dataSource && locationMember) {

          dataSource.on('changed', () => {

            this.markers$.next(dataSource.items()?.map(item => ({
              location: getByPath(item, locationMember),
              onClick: () => this.onMarkerClicked(item)
            })));            
          });

          dataSource.load();
        }
      });
  }

  public onMapMouseMove(e: MouseEvent) {
    this.mouseTarget = document.elementFromPoint(e.clientX, e.clientY);
  }

  public onMarkerClicked(item: CrudItem) {
    if (this.mouseTarget) {
      

      this.crudService.selectOptions({ item, target: this.mouseTarget, defaultEditLayout: 'primary' });
    }
  }
}
