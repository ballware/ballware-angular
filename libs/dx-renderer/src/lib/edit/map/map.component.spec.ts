import { ComponentFixture, TestBed } from '@angular/core/testing';

import { EditLayoutMapComponent } from './map.component';
import { Provider } from '@angular/core';
import { EDIT_SERVICE, SETTINGS_SERVICE, SettingsService } from '@ballware/meta-services';
import { EditLayoutItem } from '@ballware/meta-model';
import { mockedEditServiceContext } from '../../../test/editservice.spec';
import { BehaviorSubject } from 'rxjs';

describe('EditLayoutMapComponent', () => {
  let component: EditLayoutMapComponent;
  let fixture: ComponentFixture<EditLayoutMapComponent>;

  const mockedEditService = mockedEditServiceContext();
        
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ EditLayoutMapComponent ],      
      providers: [        
        {
          provide: SETTINGS_SERVICE,
          useValue: {  
            googlekey$: new BehaviorSubject<string|undefined>('unlicensed')
          }
        },
        {
          provide: EDIT_SERVICE,
          useFactory: () => mockedEditService.mock.object()
        } as Provider         
      ]
    })
    .compileComponents();
  });

  it('should create', () => {
    fixture = TestBed.createComponent(EditLayoutMapComponent);

    const layoutItem = {
      type: 'map',
      options: {
        dataMember: 'mockedmember'
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    fixture.componentRef.setInput('initialLayoutItem', layoutItem);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should apply options', () => {
    fixture = TestBed.createComponent(EditLayoutMapComponent);

    const layoutItem = {
      options: {
          dataMember: 'mockedmember',
          readonly: false,            
          visible: false
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    fixture.componentRef.setInput('initialLayoutItem', layoutItem);
    fixture.detectChanges();

    expect(component.livecycle.getOption('value')).toBe(null);    
    expect(component.livecycle.getOption('readonly')).toBe(false);
    expect(component.livecycle.getOption('visible')).toBe(false);

    component.livecycle.setOption('value', { lat: 0, lng: 0 });
    expect(component.livecycle.getOption('value')).toStrictEqual({ lat: 0, lng: 0 });

    component.livecycle.setOption('readonly', true);
    expect(component.livecycle.getOption('readonly')).toBe(true);

    component.livecycle.setOption('visible', true);
    expect(component.livecycle.getOption('visible')).toBe(true);

    expect(() => component.livecycle.getOption('undefined')).toThrowError('Unsupported option <undefined>');
    expect(() => component.livecycle.setOption('undefined', 'any value')).toThrowError('Unsupported option <undefined>');
  });
});
