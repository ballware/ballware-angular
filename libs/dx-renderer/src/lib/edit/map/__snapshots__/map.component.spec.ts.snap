// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLayoutMapComponent should create 1`] = `
<ballware-edit-map
  destroy={[Function Destroy]}
  googlekey$={[Function BehaviorSubject]}
  livecycle={[Function EditItemLivecycle]}
  map={[Function _DxMapComponent]}
  readonly={[Function Readonly]}
  settingsService={[Function Object]}
  value={[Function NullableLatLngValue]}
  visible={[Function Visible]}
>
  <div
    class="dx-texteditor-label mt-2"
    style="position: relative;"
  >
  </div><dx-map
    class="dx-map dx-widget dx-visibility-change-handler"
    style="width: 100%; height: 100%;"
  >
    <div
      class="dx-map-container"
    />
  </dx-map>
</ballware-edit-map>
`;
